chain/postgresdb/database/database.go

```diff
@@ -0,0 +1,342 @@
+type Config struct {
+	User         string
+	Password     string
+	Host         string
+	Name         string
+	MaxIdleConns int
+	MaxOpenConns int
+	DisableTLS   bool
+}
+
+// ConfigFromDSN provides support for creating a config from a DSN.
+func ConfigFromDSN(dsn string) (Config, error) {
+	conf, err := url.Parse(dsn)
+	if err != nil {
+		return Config{}, err
+	}
+	query := conf.Query()
+
+	password, _ := conf.User.Password()
+	cfg := Config{
+		User:       conf.User.Username(),
+		Password:   password,
+		Host:       conf.Host,
+		Name:       conf.Path,
+		DisableTLS: true,
+	}
+
+	if query.Has("sslmode") {
+		sslMode := query.Get("sslmode")
+		switch sslMode {
+		//nolint:goconst // Having to add constants is overkill.
+		case "disable":
+			cfg.DisableTLS = true
+		case "required":
+			cfg.DisableTLS = false
+		default:
+			return Config{}, fmt.Errorf("unsupported ssl mode %q. Expected disable or required", sslMode)
+		}
+	}
+
+	cfg.MaxIdleConns = 2
+	if query.Has("max-idle") {
+		max := query.Get("max-idle")
+		m, err := strconv.Atoi(max)
+		if err != nil {
+			return Config{}, fmt.Errorf("expected number for max-idle, got err: %w", err)
+		}
+		if m >= 0 {
+			cfg.MaxIdleConns = m
+		}
+	}
+
+	cfg.MaxOpenConns = 0
+	if query.Has("max-open") {
+		open := query.Get("max-open")
+		o, err := strconv.Atoi(open)
+		if err != nil {
+			return Config{}, fmt.Errorf("expected number for max-open, got err: %w", err)
+		}
+		if o >= 0 {
+			cfg.MaxOpenConns = o
+		}
+	}
+
+	return cfg, nil
+}
+
+// Open knows how to open a database connection based on the configuration.
+// It also performs a health check to make sure the connection is healthy.
+//
+//nolint:gocritic // There is nothing wrong with using value semantics here.
+func Open(ctx context.Context, cfg Config) (*sqlx.DB, error) {
+	sslMode := "require"
+	if cfg.DisableTLS {
+		sslMode = "disable"
+	}
+
+	q := make(url.Values)
+	q.Set("sslmode", sslMode)
+	q.Set("timezone", "utc")
+
+	u := url.URL{
+		Scheme:   "postgres",
+		User:     url.UserPassword(cfg.User, cfg.Password),
+		Host:     cfg.Host,
+		Path:     strings.ToLower(cfg.Name),
+		RawQuery: q.Encode(),
+	}
+
+	db, err := sqlx.Open("postgres", u.String())
+	if err != nil {
+		return nil, err
+	}
+	db.SetMaxIdleConns(cfg.MaxIdleConns)
+	db.SetMaxOpenConns(cfg.MaxOpenConns)
+
+	return db, StatusCheck(ctx, db)
+}
+
+// StatusCheck returns nil if it can successfully talk to the database. It
+// returns a non-nil error otherwise.
+func StatusCheck(ctx context.Context, db *sqlx.DB) error {
+	var pingError error
+
+	//nolint:gomnd // We want to have a reasonable retry period
+	t := time.NewTicker(100 * time.Millisecond)
+	defer t.Stop()
+
+check:
+	for {
+		select {
+		case <-t.C:
+			pingError = db.Ping()
+			if pingError == nil {
+				break check
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	const q = `SELECT true`
+	var tmp bool
+	return db.QueryRowContext(ctx, q).Scan(&tmp)
```

Comment:
We are not interested in the result, since we should always retrieve `true` if the connection is established and working correctly.
What's interesting here is the error returned by `Scan`, as that will tell us if we could make the query or not.